src/core_api.rs

```diff
@@ -2352,7 +2362,8 @@ impl<Twi: I2CForT1, D: DelayUs<u32>> Se050Backend<Twi, D> {
                 match parsed_obj_id {
                     ParsedObjectId::PersistentKey(obj)
                         if locations.contains(&Location::Internal)
-                            || locations.contains(&Location::External) =>
+                            || locations.contains(&Location::External)
+                                && kinds.contains(ItemsToDelete::KEYS) =>
```

Comment:
I find this entire function very hard to read.  I would suggest to extract one function that decides whether the object should be deleted, and one function that performs the actual deletion.  Also it would be good to have a comment explaining the logic.

This condition evaluates to `internal || (external && delete_keys)` if I read it correctly, but I think it should be `(internal || external) && delete_keys`, no?

Reply:
Hmm, right, that condition was indeed incorrect.

I did not split the function fully like you recommended because that would require a lot of reference passing for the counters and everything, but I did simplify it significantly to make it clearer and extracted intermediate values to make their meaning clearer.

I also moved the `if` statements from the match patterns to the match arms so what they do is more explicit and make it more visible that no case is missing.